dev/Repeater/UniformGridLayoutState.cpp

```diff
@@ -42,28 +42,71 @@ void UniformGridLayoutState::EnsureElementSize(
-            SetSize(realizedElement.DesiredSize(), layoutItemWidth, LayoutItemHeight, availableSize, stretch, orientation, minRowSpacing, minColumnSpacing, maxItemsPerLine);
+            realizedElement.Measure(CalculateAvailableSize(availableSize, orientation, stretch, maxItemsPerLine, layoutItemWidth, layoutItemHeight, minRowSpacing, minColumnSpacing));
+            SetSize(realizedElement.DesiredSize(), layoutItemWidth, layoutItemHeight, availableSize, stretch, orientation, minRowSpacing, minColumnSpacing, maxItemsPerLine);
         }
         else
         {
             // Not realized by flowlayout, so do this now!
-            if (const auto firstElement = context.GetOrCreateElementAt(0, winrt::ElementRealizationOptions::ForceCreate))
+            if (const auto firstElement = context.GetOrCreateElementAt(0, winrt::ElementRealizationOptions::ForceCreate).as<winrt::FrameworkElement>())
             {
-                firstElement.Measure(availableSize);
-                SetSize(firstElement.DesiredSize(), layoutItemWidth, LayoutItemHeight, availableSize, stretch, orientation, minRowSpacing, minColumnSpacing, maxItemsPerLine);
+                firstElement.Measure(CalculateAvailableSize(availableSize,orientation, stretch, maxItemsPerLine, layoutItemWidth, layoutItemHeight, minRowSpacing, minColumnSpacing));
+                SetSize(firstElement.DesiredSize(), layoutItemWidth, layoutItemHeight, availableSize, stretch, orientation, minRowSpacing, minColumnSpacing, maxItemsPerLine);
                 context.RecycleElement(firstElement);
             }
         }
     }
 }
 
+winrt::Size UniformGridLayoutState::CalculateAvailableSize(const winrt::Size availableSize,
+    const winrt::Orientation orientation,
+    const winrt::UniformGridLayoutItemsStretch& stretch,
+    const unsigned int maxItemsPerLine,
+    const double itemWidth,
+    const double itemHeight,
+    double minRowSpacing,
+    double minColumnSpacing)
+{
+    if (orientation == winrt::Orientation::Horizontal) {
+        if (!isnan(itemWidth)) {
+            double allowedColumnWidth = itemWidth;
+            if (stretch != winrt::UniformGridLayoutItemsStretch::None) {
+                allowedColumnWidth += CalculateExtraPixelsPerColumn(maxItemsPerLine, availableSize.Width, itemWidth, minColumnSpacing);
+            }
+            return winrt::Size{ (float)allowedColumnWidth, availableSize.Height};
+        }
+    }
+    else {
+        if (!isnan(itemHeight)) {
+            double allowedRowHeight = itemHeight;
+            if (stretch != winrt::UniformGridLayoutItemsStretch::None) {
+                allowedRowHeight += CalculateExtraPixelsPerColumn(maxItemsPerLine, availableSize.Height, itemHeight, minRowSpacing);
+            }
+            return winrt::Size{availableSize.Width, (float)itemHeight};
+        }
+    }
+    return availableSize;
+}
+
+double UniformGridLayoutState::CalculateExtraPixelsPerColumn(unsigned int maxItemsPerLine,
```

Comment:
nit: CalculateExtraPixelsInLine ? since it is row or column based on orientation.